File: cpp/solutions/day1.cpp
#include <common/common.hpp>

static std::shared_ptr<spdlog::logger> logger;

std::pair<std::vector<s32>, std::vector<s32>> getLists(
    std::vector<std::string> &lines) {
  std::vector<s32> l1(lines.size());
  std::vector<s32> l2(lines.size());

  s32 v1, v2;

  for (const auto &line : lines) {
    std::istringstream iss(line);
    // >> stops at whitespace
    if (!(iss >> v1 >> v2)) {
      std::runtime_error("Failed to extract characters!");
    } else {
      l1.push_back(v1);
      l2.push_back(v2);
    }
  }

  std::pair<std::vector<s32>, std::vector<s32>> lists(l1, l2);
  return lists;
}

s32 getDistance(const std::pair<std::vector<s32>, std::vector<s32>> &lists) {
  // Distance is calculated as absolute value of the difference
  // of the smallest values, so we just pop from the heap with our
  // formula.

  std::priority_queue<s32, std::vector<s32>, std::greater<s32>> h1;
  std::priority_queue<s32, std::vector<s32>, std::greater<s32>> h2;

  for (const auto &n : lists.first) {
    h1.push(n);
  }

  for (const auto &n : lists.second) {
    h2.push(n);
  }

  s32 distance{};
  // They're the same size so doesn't matter which one.
  while (!h1.empty()) {
    distance += abs(h1.top() - h2.top());
    h1.pop();
    h2.pop();
  }
  return distance;
}

s32 getSimilarity(std::pair<std::vector<s32>, std::vector<s32>> &lists) {
  // For part 2 of day 1, you have to calculate a "similarity score".
  // Essentially, for every number in the first list, how many times does it
  // occur in the second? We can accomplish this pretty easily with a hashmap.
  std::map<s32, s32> cnt;
  for (const auto n2 : lists.second) {
    cnt[n2] += 1;
  }
  s32 score{};
  for (const auto n1 : lists.first) {
    score += (cnt[n1] * n1);
  }
  return score;
}

void day1_solve(const std::string &file_path, bool debug) {
  logger = getLogger("Day 1", debug);

  auto lines = getLines(file_path.c_str());
  auto lists = getLists(lines);

  auto distance = getDistance(lists);
  logger->info("The distance score between the two lists is: {}", distance);
  auto similarity_score = getSimilarity(lists);
  logger->info("The similarity score calculated between the two lists is: {}",
               similarity_score);
}

File: python/solutions/day1.py
from .common import *
from collections import Counter
from heapq import *
import copy
from pathlib import Path
import logging

logger: logging.Logger
DAY: str = "day1"


def get_similarity_score(lists: tuple[list, list]) -> int:
    l1: list[int]
    l2: list[int]

    l1, l2 = lists

    cnt: Counter = Counter(l2)
    similarity_score: int = 0

    for n in l1:
        if n in cnt:
            similarity_score += n * cnt[n]

    return similarity_score


def get_distance(lists: tuple[list, list]) -> int:

    h1: list[int] = copy.deepcopy(lists[0])
    h2: list[int] = copy.deepcopy(lists[1])

    heapify(h1)
    heapify(h2)

    distance: int = 0

    for _ in range(len(h1)):
        distance += abs(heappop(h1) - heappop(h2))

    return distance


def get_lists(lines: list[str]) -> tuple[list, list]:

    l1: list[int] = []
    l2: list[int] = []

    for line in lines:
        n1, n2 = line.split()
        l1.append(int(n1))
        l2.append(int(n2))

    return (l1, l2)


def solve(file_path: Path, logging_level: int) -> None:
    global logger
    logger = get_logger(DAY, logging_level)
    lines: list[str] = read_file(file_path)

    lists: tuple[list, list] = get_lists(lines)

    distance: int = get_distance(lists)
    similarity_score: int = get_similarity_score(lists)

    logger.info(f"For part 1, the distance score is: {distance}.")
    logger.info(f"For part 2, the similarity score is: {similarity_score}.")


def main() -> None:
    file_path: str
    file_path, logging_level = quick_parse(DAY)
    solve(Path(file_path), logging_level)


if __name__ == "__main__":
    main()
